### DISCUSSION TITLE
How to do selective flushing using session?

### DISCUSSION THREAD

**Discussion by febus982** [sqlalchemy/sqlalchemy#9292] [Usage Questions] [answered]
Using SQLAlchemy 2.0
Assumed I have 2 nested models working as an aggregate:

```python
    class ParentModel(declarative_base):
        parent_model_id = Column(Integer, primary_key=True, autoincrement=True)
        name = Column(String)

        children = relationship("ChildModel", back_populates="parent", cascade="all, delete-orphan", lazy="selectin")

    class ChildModel(declarative_base):
        child_model_id = Column(Integer, primary_key=True, autoincrement=True)
        parent_model_id = Column(Integer, ForeignKey("parent_model.parent_model_id"), nullable=False)
        name = Column(String)

        parent = relationship("ParentModel", back_populates="children", lazy="selectin")
```

I'd like to do 3 isolated operations using the Session capabilities:

* get different models from the DB
* modify some of the child models
* save **only some** of the modified models in the DB, by flushing the parent model

The only way I was able to achieve this behaviour is using 2 separate transaction blocks and `expire_on_commit=False`:

```python
with Session() as session:
    m1 = session.get(ParentModel, 1)
    m2 = session.get(ParentModel, 2)

m1.children[0].name = "newName"
m2.children[0].name = "newName"

with Session() as session:
    session.add(m2)
```

This works fine, but `session.add` triggers a SELECT to get the models from the DB again, which is not very efficient. I would like to commit changes to `m2` aggregate while keeping in the session the changes done to `m1`, to eventually commit them later.

Doing it inside a single transaction would commit both the modifications. I can't `expunge` the `m2` model from the transaction because I might need to persist it later (it would become again using 2 transactions)

I tried using the `objects` parameter in flush:

```python
with Session() as session:
    m1 = session.get(ParentModel, 1)
    m2 = session.get(ParentModel, 2)

    m1.children[0].name = "newName"
    m2.children[0].name = "newName"

    # It is a requirement that write operations are done after all the changes above
    session.flush(objects=[m2])
    # Do other things, eventually flush m1
    session.expunge_all() # Expunging anything not yet flushed
```

I would have expected the flush to cascade through the relationship and produce the updates for `m2` children, but it's not working (I suspect because `m2` itself is not changed).

I wouldn't like to go down the rabbit hole of writing a procedure to traverse and flush all relationships models manually.

Is flush supposed to behave in this way?
What would the suggested approach be?

**Comment by zzzeek** [comment, marked as answer]
> The only way I was able to achieve this behaviour is using 2 separate transaction blocks and `expire_on_commit=False`:
> 
> ```python
> with Session() as session:
>     m1 = session.get(ParentModel, 1)
>     m2 = session.get(ParentModel, 2)
> 
> m1.children[0].name = "newName"
> m2.children[0].name = "newName"
> 
> with Session() as session:
>     session.add(m2)
> ```

the specific way that code is written, the expire_on_commit should not matter.

> 
> This works fine, but `session.add` triggers a SELECT to get the models from the DB again, which is not very efficient. 

I can't reproduce that, see MCVE below.

> 
> I tried using the `objects` parameter in flush:
> 

objects is not really recommended much because as noted it's very specific and can't really make good decisions about more complex object graphs.

> 
> I would have expected the flush to cascade through the relationship and produce the updates for `m2` children, but it's not working (I suspect because `m2` itself is not changed).

I can't comment for if you are using the "objects" parameter, if you are giving it a specific list of objects, the ORM doesnt really have the leeway to guess that you also meant a bunch of other objects also.  hence we don't really recommend using it for things.


> 
> Is flush supposed to behave in this way? What would the suggested approach be?

flush with a list of objects will try to flush exactly what you gave it and nothing else.  

suggested approach would be mostly what you're doing, depending on specifics you might want to use merge(), you might want to manufacture detached objects using [make_transient_to_detached](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.make_transient_to_detached) , there's a bunch of ways to manipulate state.

MCVE below shows no additional SQL being emitted during flush, no expire_on_commit needed, you'd need to provide a complete example illustrating where unwanted SQL is being emitted and we can break it down more closely.

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session
from sqlalchemy.orm import session

Base = declarative_base()


class ParentModel(Base):
    __tablename__ = "parent_model"

    parent_model_id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String)

    children = relationship(
        "ChildModel",
        back_populates="parent",
        cascade="all, delete-orphan",
        lazy="selectin",
    )


class ChildModel(Base):

    __tablename__ = "child_model"

    child_model_id = Column(Integer, primary_key=True, autoincrement=True)
    parent_model_id = Column(
        Integer, ForeignKey("parent_model.parent_model_id"), nullable=False
    )
    name = Column(String)

    parent = relationship(
        "ParentModel", back_populates="children", lazy="selectin"
    )

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)


with Session(e) as session:
    session.add_all([
        ParentModel(parent_model_id=1, children=[
            ChildModel(name='n1'),
            ChildModel(name='n2'),
            ChildModel(name='n3'),
        ]),
        ParentModel(parent_model_id=2, children=[
            ChildModel(name='n4'),
            ChildModel(name='n5'),
            ChildModel(name='n6'),
        ])
    ])
    session.commit()

print("----------------------------")

with Session(e) as session:
    m1 = session.get(ParentModel, 1)
    m2 = session.get(ParentModel, 2)

print("----------------------------")

m1.children[0].name = "newName"
m2.children[0].name = "newName"

print("----------------------------")
with Session(e) as session:
    session.add(m2)
    session.commit()

```

output:

```sql
2023-02-13 10:23:46,841 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-02-13 10:23:46,842 INFO sqlalchemy.engine.Engine INSERT INTO parent_model (parent_model_id, name) VALUES (?, ?)
2023-02-13 10:23:46,842 INFO sqlalchemy.engine.Engine [generated in 0.00011s] [(1, None), (2, None)]
2023-02-13 10:23:46,843 INFO sqlalchemy.engine.Engine INSERT INTO child_model (parent_model_id, name) VALUES (?, ?), (?, ?), (?, ?), (?, ?), (?, ?), (?, ?) RETURNING child_model_id
2023-02-13 10:23:46,843 INFO sqlalchemy.engine.Engine [generated in 0.00007s (insertmanyvalues)] (1, 'n1', 1, 'n2', 1, 'n3', 2, 'n4', 2, 'n5', 2, 'n6')
2023-02-13 10:23:46,843 INFO sqlalchemy.engine.Engine COMMIT
----------------------------
2023-02-13 10:23:46,844 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-02-13 10:23:46,845 INFO sqlalchemy.engine.Engine SELECT parent_model.parent_model_id AS parent_model_parent_model_id, parent_model.name AS parent_model_name 
FROM parent_model 
WHERE parent_model.parent_model_id = ?
2023-02-13 10:23:46,845 INFO sqlalchemy.engine.Engine [generated in 0.00012s] (1,)
2023-02-13 10:23:46,847 INFO sqlalchemy.engine.Engine SELECT child_model.parent_model_id AS child_model_parent_model_id, child_model.child_model_id AS child_model_child_model_id, child_model.name AS child_model_name 
FROM child_model 
WHERE child_model.parent_model_id IN (?)
2023-02-13 10:23:46,847 INFO sqlalchemy.engine.Engine [generated in 0.00017s] (1,)
2023-02-13 10:23:46,848 INFO sqlalchemy.engine.Engine SELECT parent_model.parent_model_id AS parent_model_parent_model_id, parent_model.name AS parent_model_name 
FROM parent_model 
WHERE parent_model.parent_model_id = ?
2023-02-13 10:23:46,848 INFO sqlalchemy.engine.Engine [cached since 0.00286s ago] (2,)
2023-02-13 10:23:46,849 INFO sqlalchemy.engine.Engine SELECT child_model.parent_model_id AS child_model_parent_model_id, child_model.child_model_id AS child_model_child_model_id, child_model.name AS child_model_name 
FROM child_model 
WHERE child_model.parent_model_id IN (?)
2023-02-13 10:23:46,849 INFO sqlalchemy.engine.Engine [cached since 0.001699s ago] (2,)
2023-02-13 10:23:46,849 INFO sqlalchemy.engine.Engine ROLLBACK
----------------------------
----------------------------
2023-02-13 10:23:46,849 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-02-13 10:23:46,850 INFO sqlalchemy.engine.Engine UPDATE child_model SET name=? WHERE child_model.child_model_id = ?
2023-02-13 10:23:46,850 INFO sqlalchemy.engine.Engine [generated in 0.00009s] ('newName', 4)
2023-02-13 10:23:46,850 INFO sqlalchemy.engine.Engine COMMIT

```

**Reply by febus982** [reply]
Thanks for the quick and detailed reply!

It turns out I was doing `session.close()` and failed to copy that part in my message (the real code is much more complex) 🤦
